src/librustc_parse/parser/item.rs

```diff
@@ -1324,49 +1324,43 @@ impl<'a> Parser<'a> {
         let id = self.parse_ident()?;
         let mut generics = self.parse_generics()?;
         generics.where_clause = self.parse_where_clause()?;
-        self.expect(&token::OpenDelim(token::Brace))?;
 
         let enum_definition = self.parse_enum_def(&generics).map_err(|e| {
             self.recover_stmt();
-            self.eat(&token::CloseDelim(token::Brace));
             e
         })?;
         Ok((id, ItemKind::Enum(enum_definition, generics), None))
     }
 
-    /// Parses the part of an enum declaration following the `{`.
+    /// Parses the part of an enum declaration starting with the `{`.
     fn parse_enum_def(&mut self, _generics: &Generics) -> PResult<'a, EnumDef> {
-        let mut variants = Vec::new();
-        // FIXME: Consider using `parse_delim_comma_seq`.
-        // We could then remove eating comma in `recover_nested_adt_item`.
-        while self.token != token::CloseDelim(token::Brace) {
-            let variant_attrs = self.parse_outer_attributes()?;
-            let vlo = self.token.span;
-
-            let vis = self.parse_visibility(FollowedByType::No)?;
-            if !self.recover_nested_adt_item(kw::Enum)? {
-                // Item already parsed, we need to skip this variant.
-                continue
+        let (variants, _) = self.parse_delim_comma_seq(token::Brace, |p| {
```

Comment:
I think it would be simpler if the closure would be moved to a function `fn parse_enum_variant(...) -> PResult<'a, Option<Variant>>` and then `self.parse_delim_comma_seq` + the rest can be moved into `parse_item_enum`.